03-document-classification/app.py

```diff
@@ -0,0 +1,230 @@
+    value: string; // the value of the key-value pair
+    explanation: string; // what this key-value pair represents or is
+}
+
+interface DocumentAnalysis {
+ evidence: Evidence[]; // a list of the various evidences that the document provides
+ document_type: string | null; // the type of document, e.g., "driver's license", "utility bill", "student id", etc.
+ legible: boolean; // if the document is legible, e.g., false if the document is blurry
+ contains_photo; // if the document contains a photo of the person to whom the document belongs
+ address: string | null; // the  person's residence if the document contains an address
+ name: string | null; // the person's name if the document mentions a name
+ extracted_information: KeyValue[]; // a list of key-value pairs extracted from the document, based on your judgement of what you think is important
+}
+
+Remember, do not respond with anything other than valid JSON.
+"""
+
+
+def timed(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        start_time = time.time()  # Record the start time
+        result = func(*args, **kwargs)  # Call the function
+        end_time = time.time()  # Record the end time
+        execution_time = end_time - start_time  # Calculate execution time
+        return (
+            result,
+            execution_time,
+        )  # Append the execution time to the original result
+
+    return wrapper
+
+
+def main():
+    st.title("Categorize documents")
+
+    with st.form("upload-files", clear_on_submit=True):
+        files = st.file_uploader(
+            "Upload documents",
+            accept_multiple_files=True,
+        )
+
+        if st.form_submit_button("Process documents", type="primary") and files:
+            process_files(files)
+            files = None
+
+
+def process_files(files):
+    results = []
+    for file in files:
+        with st.spinner(f"Processing {file.name}..."):
+            results.append(process_file(file))
+    display_results(results)
+
+
+@timed
+def process_file(file):
+    base64_image = base64.b64encode(file.read()).decode("utf-8")
+    client = get_client()
+    response = client.chat.completions.create(
+        model="gpt-4o",
+        messages=[
+            {
+                "role": "user",
+                "content": [
+                    {
+                        "type": "text",
+                        "text": PROMPT,
+                    },
+                    {
+                        "type": "image_url",
+                        "image_url": {
+                            "url": f"data:image/jpeg;base64,{base64_image}",
+                        },
+                    },
+                ],
+            }
+        ],
+        max_tokens=4_096,
+    )
+    response_string = response.choices[0].message.content
+    try:
+        response_content = json.loads(response_string)
+    except Exception:
+        # GPT4-turbo sometimes likes to wrap it with ```json {content here}```
```

Comment:
You could try [JSON mode](https://platform.openai.com/docs/guides/text-generation/json-mode)

Reply:
😮  I had no idea!